app/products/calls/connection.ts

```diff
@@ -56,7 +56,12 @@ export async function newClient(channelID: string, closeCb: () => void, setScree
         if (peer) {
             peer.destroy();
         }
-        InCallManager.stop();
+
+        // Allow time for the peer to be destroyed, which avoids the following error that can cause problems with accessing the audio system:
+        // AVAudioSession_iOS.mm:1243  Deactivating an audio session that has running I/O. All I/O should be stopped or paused prior to deactivating the audio session.
+        setTimeout(() => {
+            InCallManager.stop();
+        }, 1000);
```

Comment:
I am thinking (and hoping) we should know when it's safe to call this. The `peer` emits a `close` event we could probably use. 

Also, I know we discussed this briefly internally but are we making sure to release the streams/tracks properly in here?

https://github.com/mattermost/mattermost-mobile/blob/a56b1eebdf91a8aa379f7a0216227a858de3dae4/app/products/calls/connection.ts#L50-L54

From my extremely superficial investigation into the native dependency the `stop()` method implementation for both MediaStream and MediaStreamTrack doesn't seem to do much more than disabling them. 

There is however an additional  `release()` method that could be called as well and sounds more promising as to actually letting go of the allocated resources. I am wondering if you had a chance to look into this as well.

Also note that I feel the code I linked is slightly misplaced (my fault). To be thorough we should probably destroy the peer prior to stopping/releasing the tracks.

Reply:
Ok, looking into this here's what I've learned:
- peer's `close` event fires,  but the native code is async, so we're getting racy errors with InCallManager
- `track.release` is only needed for local tracks (like microphone), not remote. But I'm calling it for all the streams because it's a no-op if it's a remote only stream.
- `stream.release` causes Android to crash (releasing isn't supported with unified plan), but iOs doesn't complain for a reason I can't explain. So I'm not going to call it. So far it doesn't seem to be causing any problems.
- I think `peer.close` should be called after the tracks are stopped, because internally (in native land) it removes the tracks and streams from the dictionary tracking them, which means they won't be around when you try to close/release them afterwards.
- tracing `track.close` all the way into the google c++ code, it looks like closing a track is translated into a `set_enabled(false)`, which according to the comments is enough. There's not a lot of documentation about what's happening internally, and I had to stop for time. Maybe in the future if I learn the code a bit more I can figure out how the webrtc internals work.
-

Reply:
Thanks for the thorough explanation @cpoile :100: 
I think this is more than enough for now. We can always iterate further as we discover new details.